flake8_trio/visitors/visitor91x.py

```diff
@@ -186,38 +204,103 @@ def visit_FunctionDef(self, node: cst.FunctionDef) -> None:
-            # updated_node does not have a Position, so we must send original_node
-            self.check_function_exit(original_node)
-        self.restore_state(original_node)
-        return updated_node
+        if (
+            self.async_function
+            # updated_node does not have a position, so we must send original_node
+            and self.check_function_exit(original_node)
+            and self.options.autofix
+            and isinstance(updated_node.body, cst.IndentedBlock)
+        ):
+            # insert checkpoint at the end of body
+            new_body = list(updated_node.body.body)
+            new_body.append(self.checkpoint_statement())
+            indentedblock = updated_node.body.with_changes(body=new_body)
+            updated_node = updated_node.with_changes(body=indentedblock)
 
-    # error if function exits or returns with uncheckpointed statements
-    def check_function_exit(self, node: cst.Return | cst.FunctionDef):
-        for statement in self.uncheckpointed_statements:
-            self.error_91x(node, statement)
+        self.restore_state(original_node)
+        return updated_node  # noqa: R504
+
+    # error if function exit/return/yields with uncheckpointed statements
+    def check_function_exit(
+        self,
+        original_node: cst.FunctionDef | cst.Return | cst.Yield,
+    ) -> bool:
+        # Add this as a node potentially needing checkpoints only if it
+        # solely depends on whether the artificial statement is "real"
+        if self.uncheckpointed_statements == {ARTIFICIAL_STATEMENT}:
+            # function can't end in the middle of a loop body, where artificial
+            # statements are injected
+            assert not isinstance(original_node, cst.FunctionDef)
+
+            self.loop_state.nodes_needing_checkpoints.append(original_node)
+
+        # list comprehension inside the any so it doesn't short-circuit
+        return any(
+            [  # noqa: PIE802
+                self.error_91x(original_node, statement)
+                for statement in self.uncheckpointed_statements
+            ]
+        )
 
     def leave_Return(
         self, original_node: cst.Return, updated_node: cst.Return
     ) -> cst.Return:
         if not self.async_function:
             return updated_node
-        self.check_function_exit(original_node)
+        if self.check_function_exit(original_node):
+            self.add_statement = self.checkpoint_statement()
         # avoid duplicate error messages
         self.uncheckpointed_statements = set()
 
+        # return original node to avoid problems with identity equality
+        assert original_node.deep_equals(updated_node)
+        return original_node
+
+    # TODO: generate an error in these two if transforming+visiting is done in a single
+    # pass and emit-error-on-transform can be enabled/disabled. The error can't be
+    # generated in the yield/return since it doesn't know if it will be autofixed.
```

Comment:
unlikely I'll go with that design, transform+visit in single pass I'm pretty sure would guarantee that errors are output with incorrect line markers so I think you have to first transform, then revisit without transform. But would be substantial runtime reduction if you can do in single pass ofc, so will have to play around a bit with it.